src/search/landmarks/landmark_heuristic.cc

```diff
@@ -70,6 +79,42 @@ void LandmarkHeuristic::initialize(const plugins::Options &opts) {
     }
 }
 
+bool LandmarkHeuristic::landmark_graph_has_cycle_of_natural_orderings() {
+    int num_landmarks = lm_graph->get_num_landmarks();
+    vector<bool> closed(num_landmarks, false);
+    vector<bool> visited(num_landmarks, false);
+    for (auto &node : lm_graph->get_nodes()) {
+        if (depth_first_search_for_cycle_of_natural_orderings(
+                *node, closed, visited)) {
+            return true;
+        }
+    }
+    return false;
+}
+
+bool LandmarkHeuristic::depth_first_search_for_cycle_of_natural_orderings(
+    const LandmarkNode &node, vector<bool> &closed, vector<bool> &visited) {
+    int id = node.get_id();
+    if (closed[id]) {
+        return false;
+    } else if (visited[id]) {
+        // Cycle detected.
+        return true;
+    }
+
+    visited[id] = true;
+    for (auto &child : node.children) {
+        if (child.second >= EdgeType::NATURAL) {
+            if (depth_first_search_for_cycle_of_natural_orderings(
+                    *child.first, closed, visited)) {
+                return true;
+            }
+        }
+    }
+    closed[id] = true;
+    return false;
+}
```

Comment:
OK, so this is all O(N). Please ignore my previous comment; apologies. And I think the code is pretty enough; the only thing I somewhat dislike is that it would be nice to have this as a generic graph algorithm and then have the landmark code use this, but that requires some design work I wouldn't bother with at this point.

Reply:
I was thinking about a more generic implementation but decided against it exactly because I think it is beyond the scope of this issue. Should I add a TODO or open a new issue right away?

Reply:
No, I don't think it's worth it. We can leave this as is.